autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/configuration/integration/modern/listener/AutofillWebMessageListener.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.autofill.impl.configuration.integration.modern.listener
+
+import android.annotation.SuppressLint
+import androidx.annotation.CheckResult
+import androidx.collection.LruCache
+import androidx.webkit.JavaScriptReplyProxy
+import androidx.webkit.WebViewCompat
+import com.duckduckgo.autofill.api.Callback
+import com.duckduckgo.common.utils.ConflatedJob
+import java.util.*
+
+/**
+ * Base class for handling autofill web messages, which is how we communicate between JS and native code for autofill
+ *
+ * Each web message will have a unique key which is used in the JS when initiating a web message. e.g., `window.ddgGetAutofillData.postMessage()`
+ * And so each listener will declare which key they respond to.
+ *
+ * Each listener can also declare which origins they permit messages to come from. by default it will respond to all origins unless overridden.
+ *
+ * When a web message is received, there will be a `reply` object attached which is how the listener can respond back to the JS.
+ * If the listener needs to interact with the user first, it should call [storeReply] which will return it a `requestId`.
+ * This `requestId` can then be provided later to [AutofillMessagePoster] which will route the message to the correct receiver in the JS.
+ *
+ * The recommended way to declare a new web message listener is add a class which extends this abstract base class and
+ * annotate it with `@ContributesMultibinding(FragmentScope::class)`. This will then be automatically registered and unregistered
+ * when a new WebView in a tab is initialised or destroyed. See [InlineBrowserAutofill] for where this automatic registration happens.
+ */
+abstract class AutofillWebMessageListener : WebViewCompat.WebMessageListener {
+
+    abstract val key: String
+
+    /**
+     * By default, a web message listener can be sent messages from all origins. This can be overridden to restrict to specific origins.
+     */
+    open val origins: Set<String> get() = setOf("*")
+
+    lateinit var callback: Callback
+    lateinit var tabId: String
```

Comment:
Q: I know we will have one instance per tab. why do we need tabId here? wonder if this is specific of a usecase or applies to all webmessages. (maybe document if necessary)

Reply:
At the moment there's only one usecase for it. It might be something that could be moved outside somehow, but i'm going to keep it there for now in the interests of not holding the project up for it as i don't think it's an essential change.